File: sdks/golang/pkg/frame/frame.go
// Package frame implements the Krymux wire framing, byte-for-byte
// compatible with the Rust reference implementation (crates/krymux/src/frame.rs):
//
//	[type:1][flags:1][streamId:4 BE][length:4 BE]
//
// followed by the payload.
package frame

import (
	"encoding/binary"
	"errors"
	"fmt"
)

// Frame types (note: 1-based, matching the Rust FT_* constants exactly).
const (
	Hello   byte = 1 // session hello (JSON capabilities exchange)
	Open    byte = 2 // open a new stream (JSON target description)
	OpenAck byte = 3 // accept or reject a stream open (JSON verdict)
	Data    byte = 4 // stream payload bytes
	Window  byte = 5 // receive-window credit grant (4-byte BE delta)
	Close   byte = 6 // abort a stream (code/reason informational)
	Ping    byte = 7 // keepalive ping (16-byte nonce)
	Pong    byte = 8 // keepalive pong (echoes the ping nonce)
	Goaway  byte = 9 // tear down the whole session (JSON reason)
)

// Frame flags.
const (
	FlagCompressed byte = 0x01 // the DATA payload is compressed with the stream's algorithm
	FlagFIN        byte = 0x02 // end of the stream's send direction (half-close)
)

// Size limits, mirroring the Rust constants.
const (
	HeaderSize            = 10
	HardMaxFrame          = 1 << 20   // hard cap on any single frame's payload
	DefaultMaxData uint32 = 64 * 1024 // default maximum DATA payload the peer may send per frame
	MaxHello              = 8 * 1024  // maximum HELLO payload size
	MaxOpen               = 8 * 1024  // maximum OPEN payload size
	MaxControl            = 8 * 1024  // maximum OPEN_ACK / GOAWAY payload size
)

// Errors returned by ParseHeader.
var (
	ErrShortHeader = errors.New("frame: header shorter than 10 bytes")
	ErrUnknownType = errors.New("frame: unknown frame type")
)

// Header is a parsed frame header (everything before the payload).
type Header struct {
	Type     byte   // one of the frame type constants
	Flags    byte   // bitmask of Flag* flags
	StreamID uint32 // stream the frame belongs to (0 for session-level frames)
	Length   uint32 // payload length in bytes
}

// ParseHeader parses a 10-byte header, rejecting unknown frame types.
func ParseHeader(buf []byte) (Header, error) {
	if len(buf) < HeaderSize {
		return Header{}, ErrShortHeader
	}
	t := buf[0]
	if t < Hello || t > Goaway {
		return Header{}, fmt.Errorf("%w %d", ErrUnknownType, t)
	}
	return Header{
		Type:     t,
		Flags:    buf[1],
		StreamID: binary.BigEndian.Uint32(buf[2:6]),
		Length:   binary.BigEndian.Uint32(buf[6:10]),
	}, nil
}

// HeaderBytes builds the fixed 10-byte wire header for a frame.
func HeaderBytes(frameType, flags byte, streamID uint32, length uint32) [HeaderSize]byte {
	var h [HeaderSize]byte
	h[0] = frameType
	h[1] = flags
	binary.BigEndian.PutUint32(h[2:6], streamID)
	binary.BigEndian.PutUint32(h[6:10], length)
	return h
}

// Encode builds one complete wire frame (header plus payload).
func Encode(frameType, flags byte, streamID uint32, payload []byte) []byte {
	h := HeaderBytes(frameType, flags, streamID, uint32(len(payload)))
	out := make([]byte, 0, HeaderSize+len(payload))
	out = append(out, h[:]...)
	return append(out, payload...)
}

// MaxPayloadFor mirrors the receive caps the Rust reader enforces per frame
// type (a violation draws a GOAWAY). ourMaxData is the per-frame DATA cap we
// advertised in HELLO.
func MaxPayloadFor(frameType byte, ourMaxData int) int {
	switch frameType {
	case Hello:
		return MaxHello
	case Open:
		return MaxOpen
	case Window:
		return 4
	case Close:
		return 1 + 256
	case Ping, Pong:
		return 64
	case OpenAck, Goaway:
		return MaxControl
	case Data:
		return ourMaxData
	default:
		return HardMaxFrame
	}
}

File: crates/krymux/src/frame.rs
//! CMPX wire framing, byte-for-byte compatible with the Node reference
//! implementation: `[type:1][flags:1][streamId:4][length:4]` big-endian,
//! followed by the payload.

/// Frame type: session hello (JSON capabilities exchange).
pub const FT_HELLO: u8 = 1;
/// Frame type: open a new stream (JSON target description).
pub const FT_OPEN: u8 = 2;
/// Frame type: accept or reject a stream open (JSON verdict).
pub const FT_OPEN_ACK: u8 = 3;
/// Frame type: stream payload bytes.
pub const FT_DATA: u8 = 4;
/// Frame type: receive-window credit grant (4-byte delta).
pub const FT_WINDOW: u8 = 5;
/// Frame type: abort a stream (code and reason are informational).
pub const FT_CLOSE: u8 = 6;
/// Frame type: keepalive ping (16-byte nonce).
pub const FT_PING: u8 = 7;
/// Frame type: keepalive pong (echoes the ping nonce).
pub const FT_PONG: u8 = 8;
/// Frame type: tear down the whole session (JSON reason).
pub const FT_GOAWAY: u8 = 9;

/// Flag: the DATA payload is compressed with the stream's algorithm.
pub const FLAG_COMPRESSED: u8 = 0x01;
/// Flag: end of the stream's send direction (half-close).
pub const FLAG_FIN: u8 = 0x02;

/// Size of the fixed frame header, in bytes.
pub const HEADER_SIZE: usize = 10;
/// Hard cap on any single frame's payload, DATA included.
pub const HARD_MAX_FRAME: usize = 1 << 20;
/// Default maximum DATA payload the peer may send per frame.
pub const DEFAULT_MAX_DATA: u32 = 64 * 1024;
/// Maximum HELLO payload size.
pub const MAX_HELLO: usize = 8 * 1024;
/// Maximum OPEN payload size.
pub const MAX_OPEN: usize = 8 * 1024;
/// Maximum OPEN_ACK / GOAWAY payload size.
pub const MAX_CONTROL: usize = 8 * 1024;

// Wire-protocol close-code registry (see PROTOCOL.md §4.1) — documentation
// surface for interoperability; the tunnel treats any CLOSE as an abort.
/// Close code: clean end of stream.
#[allow(dead_code)]
pub const CLOSE_EOS: u8 = 0;
/// Close code: stream-level error.
pub const CLOSE_ERROR: u8 = 1;
/// Close code: upstream unreachable.
pub const CLOSE_UNREACHABLE: u8 = 3;
/// Close code: the session is going away.
pub const CLOSE_GOAWAY: u8 = 4;
/// Close code: stream cancelled by its opener.
pub const CLOSE_CANCEL: u8 = 5;

/// A parsed frame header (everything before the payload).
#[derive(Debug, Clone, Copy)]
pub struct FrameHeader {
    /// Frame type; one of the `FT_*` constants.
    pub frame_type: u8,
    /// Bitmask of `FLAG_*` flags.
    pub flags: u8,
    /// Stream the frame belongs to (0 for session-level frames).
    pub stream_id: u32,
    /// Payload length in bytes.
    pub length: u32,
}

impl FrameHeader {
    /// Parses a 10-byte header, rejecting unknown frame types.
    pub fn parse(buf: &[u8; HEADER_SIZE]) -> anyhow::Result<FrameHeader> {
        let frame_type = buf[0];
        if !(FT_HELLO..=FT_GOAWAY).contains(&frame_type) {
            anyhow::bail!("unknown frame type {}", frame_type);
        }
        Ok(FrameHeader {
            frame_type,
            flags: buf[1],
            stream_id: u32::from_be_bytes([buf[2], buf[3], buf[4], buf[5]]),
            length: u32::from_be_bytes([buf[6], buf[7], buf[8], buf[9]]),
        })
    }
}

/// Builds the fixed 10-byte wire header for a frame.
pub fn frame_header(frame_type: u8, flags: u8, stream_id: u32, length: u32) -> [u8; HEADER_SIZE] {
    let mut h = [0u8; HEADER_SIZE];
    h[0] = frame_type;
    h[1] = flags;
    h[2..6].copy_from_slice(&stream_id.to_be_bytes());
    h[6..10].copy_from_slice(&length.to_be_bytes());
    h
}

/// Encodes a header plus payload into one complete wire frame.
pub fn encode_frame(frame_type: u8, flags: u8, stream_id: u32, payload: &[u8]) -> Vec<u8> {
    debug_assert!(payload.len() <= HARD_MAX_FRAME);
    let mut out = Vec::with_capacity(HEADER_SIZE + payload.len());
    out.extend_from_slice(&frame_header(
        frame_type,
        flags,
        stream_id,
        payload.len() as u32,
    ));
    out.extend_from_slice(payload);
    out
}
